### PR TITLE
Stdlib: added Printf.ikbprintf

### LINKED ISSUE #7110: Missing Printf.ikbprintf
**Original bug ID:** 7110
**Reporter:** berke.durak
**Status:** acknowledged (set by @damiendoligez on 2017-04-13T11:19:14Z)
**Resolution:** open
**Priority:** low
**Severity:** feature
**Platform:** Unix
**OS:** Linux
**OS Version:** 3.17
**Version:** 4.01.0
**Category:** standard library

## Bug description

The Printf module provides an ikfprintf function that takes a continuation and ignores the format:

    val ikfprintf :
      (out_channel -> 'a) ->
      out_channel -> ('b, out_channel, unit, 'a) format4 -> 'b

The same should be provided for buffers.

    val ikbprintf :
      (Buffer.t -> 'a) ->
      Buffer.t -> ('b, Buffer.t, unit, 'a) format4 -> 'b

## Steps to reproduce

N/A

## Additional information

I noticed the ikbprintf function was missing while updating a simple logging function to provide a "log, then raise Failure exception" functionality.

The logging function was initially written with channels, but was later updated to use Buffer formatters to support logging to multiple channels.

### PR DESCRIPTION
**Pull request by muskangarg21**
fixes #7110 
@Octachron please review. Thanks

### CODE REVIEW

**Comment by gasche** [comment]
The code is exactly the same as `ikfprintf`, and the type is also the same. I didn't expect this (it is surprising), but I think that it means that `ikfprintf` is in fact generic enough to be used for "ignoring printers" whatever their accumulator type is (`out_channel` for `fprintf`, `Buffer.t` for `bprintf`).

This generality was not initially present (in released versions) when the original issue #7110, it was added by #201 at around the same time. (I reviewed this generalizing change but didn't realize at the time that it applied to `i{k,}bprintf`.)

I think that there are two choices we could make:
1. have `ibprintf` and `ikbprintf` be documented as aliases of `ifprintf` and `ikfprintf` (the point is to let people change code by just adding `i`, even if the other function would have done the job)
2. not adding new functions, but documenting in the existing `i[k]fprintf` that it also works for the buffer version

Personally I think that (1) is the nicer option for users. This is an extension of the work currently proposed in this PR, except that:
- you could/should just do `let ikbprintf = ikfprintf`, instead of re-expanding to take arguments
- the documentation should point out that `ikfprintf` and `ikbprintf` are general synonyms of each other (but it's fine to keep the documentation you already include, which describes (correctly) the relation between `kbprintf` and `ikbprintf`)
- I think we should also add `ibprintf`, so that the `{i,}k{b,f}printf` functions and the `{i,}{b,f}printf` functions have exactly the same structure

**Comment by muskangarg21** [comment]
@gasche is this what you had in mind ?

**Comment by gasche** [comment]
This looks good to me, thanks! (I think it would be nice to also have someone else's opinion on the matter.)

The @since tags should not be "@since 4.10", but "@since 4.11", given that they will only be available in the next release.

**Comment by muskangarg21** [comment]
> This looks good to me, thanks! (I think it would be nice to also have someone else's opinion on the matter.)
> 
> The @SInCE tags should not be "@SInCE 4.10", but "@SInCE 4.11", given that they will only be available in the next release.

Okay, I ll update that.
@Octachron your views on this ?

**Comment by yallop** [comment]
> (I think it would be nice to also have someone else's opinion on the matter.)

My preference would be to restrict the type of `ikbprintf` so that it's the same as the type of `kbprintf` (and similarly for `ibprintf`/`bprintf`).  More general types are not always better: there's no advantage in being able to write `ibprintf stdout "%d" 3` (almost certainly a mistake), for example, which the current code allows.

**Comment by Octachron** [comment]
I tend to agree with @yallop, it is probably better to keep the specific type for`ikbprintf` to make sure that `ikbrintf` and `kbprintf` can always be swapped for each other in any context.

I also think it would be nice to add a `ibprintf` function for the sake of symmetry.

File: stdlib/printf.ml [hunk 1]
```diff
@@ -22,10 +22,13 @@ let kbprintf k b (Format (fmt, _)) =
   make_printf (fun acc -> bufput_acc b acc; k b) End_of_acc fmt
 let ikfprintf k oc (Format (fmt, _)) =
   make_iprintf k oc fmt
+let ikbprintf k bc (Format (fmt, _)) =
+  make_iprintf k bc fmt
```
Comment (Octachron) [inline stdlib/printf.ml]: You can keep the equality ` let ikbprintf = ikfprintf` here, even if we do not expose the equality in the interface.

File: stdlib/printf.mli [hunk 2]
```diff
@@ -147,6 +147,12 @@ val ifprintf : 'b -> ('a, 'b, 'c, unit) format4 -> 'a
     @since 3.10.0
 *)
 
+val ibprintf : Buffer.t -> ('a, Buffer.t, 'c, unit) format4 -> 'a
```
Comment (Octachron) [inline stdlib/printf.mli]: It is better to use the `format` type (and not `format4`) to get exactly the same type as `bprintf`.

File: stdlib/printf.mli [hunk 3]
```diff
@@ -175,6 +181,13 @@ val kbprintf : (Buffer.t -> 'd) -> Buffer.t ->
    @since 3.10.0
 *)
 
+val ikbprintf : (Buffer.t -> 'd) -> Buffer.t ->
+               ('a, Buffer.t, 'c, 'd) format4 -> 'a
```
Comment (Octachron) [inline stdlib/printf.mli]: Same as above, we want to restrict `'c` to `unit` to keep the exact same type as `kbprintf`.

**Comment by Octachron** [comment]
Merged, thanks for the contribution!

**Pull request merged**